pmem: mem[cpy|set] optimized for SSE2/AVX/AVX512F

marcinslusarz opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/2326)
<!-- Reviewable:end -->

marcinslusarz commented:
pmem implementation is ready. I have to add more tests (or actually refactor existing ones to test non-default ISA implementation)

marcinslusarz commented:
AVX512F is not wired up on Windows, because the version of VS we use doesn't support it. 
We can wire up it later.

GBuella commented:
I argue the SSE version is not needed.
It is just pointless to maintain that code.
Do we need to spend workhours for the case of someone using at lest 8 years old x86 CPU in their datacenter with non-volatile memory? Is it even possible in practice to use the modern NV hardware implementations we mainly target, with such old CPUs?

marcinslusarz commented:
I get your point and I don't have a strong opinion about that, but I think having SSE2 version in tree may be useful for comparisons and experimentation.

GBuella commented:
OK, probably you are the one who is going to maintain it anyways.

krzycz commented:
I agree with @Gbuella. The platforms equipped with Intel NVDIMMs should have AVX available, so SSE2 variant is probably not needed.
Also, I think it would get much more complicated if we support other architectures at some point. (i.e AARCH64).

---

Reviewed 11 of 37 files at r1.
Review status: 9 of 37 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/libpmem/pmem.c, line 871 at r2](https://reviewable.io:443/reviews/pmem/nvml/2326#-Ky1As4YKMQpEfzKAhxT:-Ky1As4ZjMsAMSBSOhGK:b7gwgf) ([raw file](https://github.com/pmem/nvml/blob/af6868997e41df694e72f31f8ddd39e81e7fb08c/src/libpmem/pmem.c#L871)):*
> ```C
> 		return dest;
> 
> 	if (len < Movnt_threshold) {
> ```

Perhaps the name of this variable is a bit misleading, but I think the point was to set the threshold at which MOVNT variant of memcpy is faster than glibc memcpy+pmem_flush.
I assume glibc provides probably the best memcpy implementation for given architecture/transfer size (I mean the best average) and the only drawback (in case of PMEM) is that we never know if it uses temporal or non-temporal stores, so we need to flush CPU caches each time glibc variant is used.

Now, we never fall back to glibc implementation, which is fine, but maybe we need another variable that would force libpmem to use glibc memcpy, instead of our temporal variants.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2326#-:-Ky1KIwVurl4_xzdMlX9:b-ccyxrc)*
<!-- Sent from Reviewable.io -->

pbalcer commented on src/libpmem/pmem.h:
```diff
@@ -42,6 +42,45 @@ void pmem_init(void);
 
 int is_pmem_detect(const void *addr, size_t len);
 
+#ifndef AVX512F_AVAILABLE
```
Why do we need to expose those functions? I don't see them being used outside pmem.c module.

marcinslusarz replied:
What do you mean? We are not exposing them.

pbalcer commented on src/libpmem/Makefile:
```diff
@@ -48,8 +48,47 @@ SOURCE =\
+	memset_t_sse2.c
+
+AVX512F_PROG="\#include <immintrin.h>\n\#include <stdint.h>\nint main(){ uint64_t v[8]; __m512i zmm0 = _mm512_loadu_si512((__m512i *)&v); return 0;}"
+AVX512F_AVAILABLE := $(shell printf $(AVX512F_PROG) |\
```
$(call check_flag, -mavx512f)

marcinslusarz replied:
Unfortunately this is not enough. With older compilers -mavx512f works, but the first time we use anything from immintrin.h they fail.

marcinslusarz commented:
@krzycz, This is not only about availability of AVX, but also about choosing the right mode for an application. If for some reason AVX instructions would negatively affect application's performance, having SSE2 version might be useful.

 This is pretty self-contained code. Other architectures support will be in different files. 

Should we remove clflush support because no one will use it with Intel NVDIMMs?

Should we compile whole NVML with -mavx?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2326#-:-Ky1TKOMDUP8ee_ZC6v_:bqgeqqr)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
@pbalcer I could use macro magic to make the code shorter, but I wanted to avoid it, because macros usually obfuscate the code.
We need both variants because I want us to have control over fallback in pmem_memcpy.
As a next PR I would like to expose both non-temporal and temporal versions as separate library entry points, because their performance depend on usage pattern.

pbalcer commented:
@marcinslusarz Making the library work on platforms that won't support NVDIMMs is different than making it fast on such platforms. One is useful for development/prototyping, the other not so much.

I'm opposed to adding any new memcpy variants to libpmem. If we can't figure out the best heuristic to chose which implementation to use, how would a user do it? Or differently, can we explain in the manpage when to use a given implementation? Because if not, than adding that API is IMHO misguided.

Why do you need to control the memcpy fallback? Is the performance noticeably better? That is A LOT of potentially unnecessary copy/pasted code.

marcinslusarz commented:
I'd understand the opposition to merging sse2 variant if it would be wildly different or required changes anywhere outside. But implementation is self-contained and very similar do the other variants, just *simpler*.

We can figure out what is the best, just not from INSIDE of libpmem.

I want to have full control over memcpy implementation to not wonder if libc in this particular version and with particular size uses something other than regular stores. If libc will start using NT instructions at some threshold in some particular version (or even worse, only in one distribution!) this may break performance of applications relying on temporal copies. There are other libc's which may have different implementation of memcpy (think at least about glibc, bionic, BSD libc and MSVCRT).

krzycz commented:
Ref: pmem/issues#656

GBuella commented on src/libpmem/memcpy/memcpy_sse2.h:
```diff
@@ -0,0 +1,103 @@
+
+		_mm_storeu_si128((__m128i *)dest, xmm0);
+		_mm_storeu_si128((__m128i *)(dest + len - 16), xmm1);
+	} else if (len > 8) {
```
How about:
```
else {
   while (len > 0) {
      *dst++ = *src++;
      --len;
   }
}
```

marcinslusarz replied:
My version does 2 loads and 2 stores. Your version can do up to 16 loads and 16 stores.

GBuella replied:
>> My version does 2 loads and 2 stores. Your version can do up to 16 loads and 16 stores.

Seriously, this part is reached following a function call, and another function call with an address loaded from a variable, and at least a dozen or so branches. If the users data is not aligned to contain only full cachelines, well then this doesn't anyways, ah, anyways nevermind

marcinslusarz replied:
I'm not sure what is your point here, but it's the same trick libc uses.

Operations on registers are extremely fast comparing to loads & stores from/to memory, so it doesn't matter much how many branches are here.

I tried to optimize this code by doing as little branches as possible, but benchmarking proved that it doesn't matter.

GBuella commented on src/libpmem/memset/memset_sse2.h:
```diff
@@ -0,0 +1,102 @@
+
+		_mm_storeu_si128((__m128i *)(dest + 0), xmm);
+		_mm_storeu_si128((__m128i *)(dest + len - 16), xmm);
+	} else if (len > 8) {
```
This might be unnecessary as well, how about:
```
else {
   memset(dest, c, len);
}
```

This would also make testing a bit easier, I mean, fewer code paths to cover.

marcinslusarz replied:
Strictly speaking nothing in this patch set is necessary, so I don't think it's a valid argument.

This function is fully covered by tests.

GBuella commented on src/libpmem/memset/memset_sse2.h:
```diff
@@ -0,0 +1,102 @@
+		_mm_storeu_si128((__m128i *)(dest + len - 16), xmm);
+	} else if (len > 32) {
+		/* 33..48 */
+		__m128i xmm = _mm_set1_epi8((char)c);
```
Why not hoist this `xmm = ` to before the first `if`, as is done in `memset_mov_avx`?

marcinslusarz replied:
You are comparing to function with different purpose. `memset_small_sse2` should be compared to `memset_small_avx`.

Moving this variable before first `if` would compute it in cases when it's not needed.

pbalcer commented:
@marcinslusarz But by having, and thereby (as ar as I understand) publicly exposing, SSE2 implementation we are effectively making unportable, ISA-specific, entry points.

Apart from the copy/pasted nature of the code, I'm personally not opposed to merging how many various memcpy implementations you want, but I think we need strong data points that show a scenarios in which using something different than the best available instruction set is a better option.

My opinion is that if an application requires this level of optimization, and isn't satisfied by the default memcpy, then it should probably link with modified version of libpmem that does expose those hardware-specific features.

marcinslusarz commented:
You misunderstood what I want to do. I don't want to expose different implementations of memcpy based on ISA.

I want only 2 additional functions - one which always use temporal instructions and one which will always non-temporal instructions. That's it.

krzycz commented:
Reviewed 4 of 37 files at r1, 1 of 2 files at r3, 44 of 105 files at r5, 27 of 70 files at r6, 5 of 11 files at r7.
Review status: 88 of 171 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/libpmem/memcpy/memcpy_t_avx.c, line 230 at r3](https://reviewable.io:443/reviews/pmem/nvml/2326#-Ky1Xb7J1A4M8lUS-AP7:-Ky1Xb7J1A4M8lUS-AP8:bh0dyla) ([raw file](https://github.com/pmem/nvml/blob/0943cffd8352e01f3b3ab2ca8c196838f7582f84/src/libpmem/memcpy/memcpy_t_avx.c#L230)):*
> ```C
> 	else
> 		memmove_mov_avx_bw(dest, src, len);
> 	_mm256_zeroupper();
> ```

what is it for? a comment would be helpful

---

*[src/test/pmem_movnt_align/TEST0, line 61 at r7](https://reviewable.io:443/reviews/pmem/nvml/2326#-KyH7m3HV9vsChfHNy8_:-KyH7m3HV9vsChfHNy8a:bbkeqkv) ([raw file](https://github.com/pmem/nvml/blob/8ca85ee042574753ebc5f736db185223136eef86/src/test/pmem_movnt_align/TEST0#L61)):*
> ```
> function test_all() {
> 	unset PMEM_MOVNT_THRESHOLD
> 	test C 0
> ```

If the second argument is the log file index, then why 0..3 occurs twice?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2326)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 7 of 37 files at r1, 1 of 1 files at r2, 1 of 2 files at r3, 1 of 1 files at r4, 15 of 105 files at r5, 38 of 70 files at r6, 6 of 11 files at r7.
Review status: 157 of 171 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2326)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Review status: 157 of 171 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/libpmem/pmem.c, line 871 at r2](https://reviewable.io:443/reviews/pmem/nvml/2326#-Ky1As4YKMQpEfzKAhxT:-KyHIpurxiatRLteqMQm:b9b0t18) ([raw file](https://github.com/pmem/nvml/blob/af6868997e41df694e72f31f8ddd39e81e7fb08c/src/libpmem/pmem.c#L871)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Perhaps the name of this variable is a bit misleading, but I think the point was to set the threshold at which MOVNT variant of memcpy is faster than glibc memcpy+pmem_flush.
I assume glibc provides probably the best memcpy implementation for given architecture/transfer size (I mean the best average) and the only drawback (in case of PMEM) is that we never know if it uses temporal or non-temporal stores, so we need to flush CPU caches each time glibc variant is used.

Now, we never fall back to glibc implementation, which is fine, but maybe we need another variable that would force libpmem to use glibc memcpy, instead of our temporal variants.

</blockquote></details>

We already have something like that - PMEM_NO_MOVNT. 

---

*[src/libpmem/memcpy/memcpy_t_avx.c, line 230 at r3](https://reviewable.io:443/reviews/pmem/nvml/2326#-Ky1Xb7J1A4M8lUS-AP7:-KyH9y2QSRkLvO2cwOJG:b1s1w4e) ([raw file](https://github.com/pmem/nvml/blob/0943cffd8352e01f3b3ab2ca8c196838f7582f84/src/libpmem/memcpy/memcpy_t_avx.c#L230)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

what is it for? a comment would be helpful
</blockquote></details>

It has 2 purposes:
- optimization (google VZEROUPPER instruction for details)
- dealing with Valgrind (see comment in memcpy_nt_avx.c)

Any idea how to reference this information everywhere _mm256_zeroupper is used without copy/pasting it? 

---

*[src/test/pmem_movnt_align/TEST0, line 61 at r7](https://reviewable.io:443/reviews/pmem/nvml/2326#-KyH7m3HV9vsChfHNy8_:-KyHIBrhNE5YpjJtgPvX:by39281) ([raw file](https://github.com/pmem/nvml/blob/8ca85ee042574753ebc5f736db185223136eef86/src/test/pmem_movnt_align/TEST0#L61)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

If the second argument is the log file index, then why 0..3 occurs twice?
</blockquote></details>

Because first 4 match files can handle both cases.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2326)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: 157 of 171 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/libpmem/memcpy/memcpy_t_avx.c, line 230 at r3](https://reviewable.io:443/reviews/pmem/nvml/2326#-Ky1Xb7J1A4M8lUS-AP7:-KyIB8gGNc2NBTi3gN8S:bf5v97m) ([raw file](https://github.com/pmem/nvml/blob/0943cffd8352e01f3b3ab2ca8c196838f7582f84/src/libpmem/memcpy/memcpy_t_avx.c#L230)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

It has 2 purposes:
- optimization (google VZEROUPPER instruction for details)
- dealing with Valgrind (see comment in memcpy_nt_avx.c)

Any idea how to reference this information everywhere _mm256_zeroupper is used without copy/pasting it? 
</blockquote></details>

One idea is to create a tiny wrapper function in a header, just to lead everyone to towards the explanation, e.g.:
```
static inline void call_it_post_avx_or_something(void)
{
   /* the long explanation goes here */
   _mm256_zeroupper();
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2326)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: 157 of 171 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/libpmem/memcpy/memcpy_sse2.h, line 75 at r6](https://reviewable.io:443/reviews/pmem/nvml/2326#-Ky0Wa6a_mkPOJ-WlMh--r6-75:-KyIGrQeLUDhFD1qlIV2:b-as2bsw) ([raw file](https://github.com/pmem/nvml/blob/a849ff777679d9f30ca2e646b9e85c7a70c8968d/src/libpmem/memcpy/memcpy_sse2.h#L75)):*
OK, sorry, a shorter version:

>> I tried to optimize this code by doing as little branches as possible, but benchmarking proved that it doesn't matter.

So, those benchmarks proved, that these 20 lines of code (needing 5 test cases) is better than a simple loop?

---

*[src/libpmem/memcpy/memcpy_t_avx.c, line 40 at r7](https://reviewable.io:443/reviews/pmem/nvml/2326#-KyIDNXT7zYzpfpahOi_:-KyIDNXT7zYzpfpahOia:b9hhs2) ([raw file](https://github.com/pmem/nvml/blob/8ca85ee042574753ebc5f736db185223136eef86/src/libpmem/memcpy/memcpy_t_avx.c#L40)):*
> ```C
> #include "pmem.h"
> 
> static inline void
> ```

The `inline` keyword is meaningless noise here, please remove.

---

*[src/libpmem/memset/memset_sse2.h, line 59 at r6](https://reviewable.io:443/reviews/pmem/nvml/2326#-Ky0Wa6a_mkPOJ-WlMh8-r6-59:-KyIG4eDnncnUQKGKERb:b-wwwlai) ([raw file](https://github.com/pmem/nvml/blob/a849ff777679d9f30ca2e646b9e85c7a70c8968d/src/libpmem/memset/memset_sse2.h#L59)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

You are comparing to function with different purpose. `memset_small_sse2` should be compared to `memset_small_avx`.

Moving this variable before first `if` would compute it in cases when it's not needed.
</blockquote></details>

Even in that case, as I recall, the microcodes for doing are just going to be executed simultaneously with the other instructions going on here. And if Mr. branch predictor predicts that `len > 16`, it might just "compute" it before (or in paralell) the branches anyways.
Unless your benchmark shows otherwise.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2326)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: 157 of 171 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/libpmem/memcpy/memcpy_sse2.h, line 75 at r6](https://reviewable.io:443/reviews/pmem/nvml/2326#-Ky0Wa6a_mkPOJ-WlMh--r6-75:-KyIOa80-Efl3ZBchBpk:b-rkuiid) ([raw file](https://github.com/pmem/nvml/blob/a849ff777679d9f30ca2e646b9e85c7a70c8968d/src/libpmem/memcpy/memcpy_sse2.h#L75)):*
<details><summary><i>Previously, GBuella (Gabor Buella) wrote…</i></summary><blockquote>

OK, sorry, a shorter version:

>> I tried to optimize this code by doing as little branches as possible, but benchmarking proved that it doesn't matter.

So, those benchmarks proved, that these 20 lines of code (needing 5 test cases) is better than a simple loop?

</blockquote></details>

I mean two simple loops, as this needs to work backwards as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2326)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 14 of 37 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[src/libpmem/pmem.c, line 871 at r2](https://reviewable.io:443/reviews/pmem/nvml/2326#-Ky1As4YKMQpEfzKAhxT:-Kz0WNM54F298FCaWsll:byr4fl6) ([raw file](https://github.com/pmem/nvml/blob/af6868997e41df694e72f31f8ddd39e81e7fb08c/src/libpmem/pmem.c#L871)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

We already have something like that - PMEM_NO_MOVNT. 
</blockquote></details>

Right!

---

*[src/libpmem/memcpy/memcpy_nt_avx.c, line 217 at r7](https://reviewable.io:443/reviews/pmem/nvml/2326#-Kz0QC6H9T9POUod3X5h:-Kz0QC6H9T9POUod3X5i:bxj7nj0) ([raw file](https://github.com/pmem/nvml/blob/8ca85ee042574753ebc5f736db185223136eef86/src/libpmem/memcpy/memcpy_nt_avx.c#L217)):*
> ```C
> 	}
> 
> 	/* There's no point in using more than 1 nt store for 1 cache line. */
> ```

I wonder if there is any performance benefit of using NT stores for anything smaller than a cache line (64B). Yes, you don't need to flush CPU cache, but I think most of the cost is in cache miss and NT store probably does not help  much. 
IOW, if we do a data transfer that is a multiply of 64B (and addresses are well-aligned), we can use NT store instructions (128/256/512-bit - whatever is available), and we know it outperforms the temporal stores in most cases, but for anything smaller than 64B (including the unaligned head/tail), we could do regular stores followed by CLWB/CLFLUSHOPT. An SFENCE is required anyway.

Do you have any data showing that NT-stores outperform temporal stores for such small data transfers?
If not, this code could be significantly simplified, as all the functions like 1x32b/1x16b/1x8b/1x4b could be eliminated.

---

*[src/libpmem/memset/memset_avx.h, line 65 at r7](https://reviewable.io:443/reviews/pmem/nvml/2326#-Kz0TqVuk5k3RASZS0JM:-Kz0TqVuk5k3RASZS0JN:b43hrfo) ([raw file](https://github.com/pmem/nvml/blob/8ca85ee042574753ebc5f736db185223136eef86/src/libpmem/memset/memset_avx.h#L65)):*
> ```objc
> 		/* 9..16 */
> 		uint64_t d;
> 		memset(&d, c, 8);
> ```

At the first glance, this code is pretty confusing, as you do the memset(..., 8) first to prepare the data, then two stores of size 8 using this data. One may ask - is it really better than memset(..., 16)?
I know it relies on some optimizations made by the compiler, but IMHO if you're doing this trick to make sure the compiler would generate some specific sequence of instructions, you should better use inline assembly. Otherwise, this code may be optimal for some compilers and suboptimal for others. Needless to say that a comment explaining why this code is expected to be faster would be helpful.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/2326)*
<!-- Sent from Reviewable.io -->

RobertCElliott commented on src/libpmem/pmem.c:
```diff
@@ -1225,6 +1018,51 @@ pmem_get_cpuinfo(void)
+#if AVX_AVAILABLE
+			LOG(3, "avx supported");
+
+			char *e = os_getenv("PMEM_NO_AVX");
```
Rather than creating a bunch of independent PMEM_* environment variables, consider adding one PMEM_TUNABLES variable modeled after GLIBC_TUNABLES (https://www.gnu.org/software/libc/manual/html_node/Tunables.html).  Present a consistent user experience for controlling similar things.  Current glibc memcpy() related tunables include hwcaps (with flags including allowing AVX, AVX2, and AVX512 usage), x86_data_cache_size, x86_shared_cache_size, and x86_non_temporal_threshold.

marcinslusarz replied:
That's a very good idea. We have something similar in pmemobj, called [pmemobj_ctl_[set|get] / PMEMOBJ_CONF](http://pmem.io/pmdk/manpages/linux/master/libpmemobj/pmemobj_ctl_get.3). Porting it to libpmem is a few days of work (code, tests, documentation), which currently I can't spend time on. In order not to forget about it I created this feature request - pmem/issues#786.

marcinslusarz commented:
Review status: 85 of 219 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/libpmem/x86_64/memcpy/memcpy_sse2.h, line 75 at r6](https://reviewable.io:443/reviews/pmem/pmdk/2326#-Ky0Wa6a_mkPOJ-WlMh--r6-75:-L4uBcpPm1eTpp8U34O0:b-zb00du) ([raw file](https://github.com/pmem/pmdk/blob/a849ff777679d9f30ca2e646b9e85c7a70c8968d/src/libpmem/x86_64/memcpy/memcpy_sse2.h#L75)):*
<details><summary><i>Previously, GBuella (Gabor Buella) wrote…</i></summary><blockquote>

I mean two simple loops, as this needs to work backwards as well.
</blockquote></details>

This optimization may not matter now, but I think it will once EADR (IOW no flushing) will be available.

---

*[src/libpmem/x86_64/memset/memset_sse2.h, line 59 at r6](https://reviewable.io:443/reviews/pmem/pmdk/2326#-Ky0Wa6a_mkPOJ-WlMh8-r6-59:-L4u8o6EnEL3ieuPhE7e:b-njoi58) ([raw file](https://github.com/pmem/pmdk/blob/a849ff777679d9f30ca2e646b9e85c7a70c8968d/src/libpmem/x86_64/memset/memset_sse2.h#L59)):*
<details><summary><i>Previously, GBuella (Gabor Buella) wrote…</i></summary><blockquote>

Even in that case, as I recall, the microcodes for doing are just going to be executed simultaneously with the other instructions going on here. And if Mr. branch predictor predicts that `len > 16`, it might just "compute" it before (or in paralell) the branches anyways.
Unless your benchmark shows otherwise.
</blockquote></details>

I changed this code to essentially do what you suggested.

---

*[src/libpmem/memset/memset_avx.h, line 65 at r7](https://reviewable.io:443/reviews/pmem/pmdk/2326#-Kz0TqVuk5k3RASZS0JM:-L-b2NBwm6wPaCuqLVVX:b-rb1bzx) ([raw file](https://github.com/pmem/pmdk/blob/8ca85ee042574753ebc5f736db185223136eef86/src/libpmem/memset/memset_avx.h#L65)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

At the first glance, this code is pretty confusing, as you do the memset(..., 8) first to prepare the data, then two stores of size 8 using this data. One may ask - is it really better than memset(..., 16)?
I know it relies on some optimizations made by the compiler, but IMHO if you're doing this trick to make sure the compiler would generate some specific sequence of instructions, you should better use inline assembly. Otherwise, this code may be optimal for some compilers and suboptimal for others. Needless to say that a comment explaining why this code is expected to be faster would be helpful.
</blockquote></details>

I changed this code significantly.

---

*[src/libpmem/memcpy/memcpy_nt_avx.c, line 217 at r7](https://reviewable.io:443/reviews/pmem/pmdk/2326#-Kz0QC6H9T9POUod3X5h:-L-b2alNja2B-J5kxRk8:bknoyve) ([raw file](https://github.com/pmem/pmdk/blob/8ca85ee042574753ebc5f736db185223136eef86/src/libpmem/memcpy/memcpy_nt_avx.c#L217)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

I wonder if there is any performance benefit of using NT stores for anything smaller than a cache line (64B). Yes, you don't need to flush CPU cache, but I think most of the cost is in cache miss and NT store probably does not help  much. 
IOW, if we do a data transfer that is a multiply of 64B (and addresses are well-aligned), we can use NT store instructions (128/256/512-bit - whatever is available), and we know it outperforms the temporal stores in most cases, but for anything smaller than 64B (including the unaligned head/tail), we could do regular stores followed by CLWB/CLFLUSHOPT. An SFENCE is required anyway.

Do you have any data showing that NT-stores outperform temporal stores for such small data transfers?
If not, this code could be significantly simplified, as all the functions like 1x32b/1x16b/1x8b/1x4b could be eliminated.

</blockquote></details>

I don't have the numbers. I optimized this code slightly.

I haven't verified it, but I think you could call memcpy_nodrain twice - for the 1st 32 bytes and for the 2nd 32 bytes and still benefit from write-combining.

---

*[src/libpmem/memcpy/memcpy_t_avx.c, line 230 at r3](https://reviewable.io:443/reviews/pmem/pmdk/2326#-Ky1Xb7J1A4M8lUS-AP7:-L-b2FaVvMcbX4jN5IMe:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/0943cffd8352e01f3b3ab2ca8c196838f7582f84/src/libpmem/memcpy/memcpy_t_avx.c#L230)):*
<details><summary><i>Previously, GBuella (Gabor Buella) wrote…</i></summary><blockquote>

One idea is to create a tiny wrapper function in a header, just to lead everyone to towards the explanation, e.g.:
```
static inline void call_it_post_avx_or_something(void)
{
   /* the long explanation goes here */
   _mm256_zeroupper();
}
```
</blockquote></details>

Done.

---

*[src/libpmem/memcpy/memcpy_t_avx.c, line 40 at r7](https://reviewable.io:443/reviews/pmem/pmdk/2326#-KyIDNXT7zYzpfpahOi_:-L-b2JDscCgjzdGNXR0Z:b-1kk2qh) ([raw file](https://github.com/pmem/pmdk/blob/8ca85ee042574753ebc5f736db185223136eef86/src/libpmem/memcpy/memcpy_t_avx.c#L40)):*
<details><summary><i>Previously, GBuella (Gabor Buella) wrote…</i></summary><blockquote>

The `inline` keyword is meaningless noise here, please remove.
</blockquote></details>

Actually, it was not enough. I changed it to force_inline.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2326)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
In the latest version of this PR I splitted libpmem into architecture-independent and architecture-dependent parts. Architecture-dependent code now lives in per-architecture directories. In order to test my changes, I managed to build and test PMDK on ARM64 under qemu. With some small changes (disabling libart build on ARM) almost all tests succeeded, with an exception of few which timed out, likely caused by software emulation of ARM CPU.

krzycz commented:
Reviewed 1 of 16 files at r8, 24 of 180 files at r9, 32 of 178 files at r10.
Review status: 140 of 219 files reviewed at latest revision, 21 unresolved discussions.

---

*[src/common/util.h, line 126 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4ud9xDeYE-IxviYxXy:-L4ud9xDeYE-IxviYxXz:b5uzpsw) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/common/util.h#L126)):*
> ```objc
> 
> #ifdef _MSC_VER
> #define force_inline inline
> ```

AFAIR, there is something like "_forceinline" in VC++

---

*[src/common/util.h, line 161 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4ud_Z4t0mO_myJEHfQ:-L4ud_Z4t0mO_myJEHfR:b-f0mnlp) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/common/util.h#L161)):*
> ```objc
> single_bit_set(uint64_t v)
> {
> 	return v && !(v & (v - 1));
> ```

1) util_is_pow2() ?
2) `return (util_popcount(v) == 1)` ?

---

*[src/examples/libpmemcto/Makefile, line 44 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4udv_Hb0xMlkFTF4Ke:-L4udv_Hb0xMlkFTF4Kf:b-ngco9r) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/examples/libpmemcto/Makefile#L44)):*
> ```Makefile
> DIRS = life
> 
> ifeq ($(ARCH), x86_64)
> ```

any comment why?
`# libart examples use x86 inline assembly ...`

---

*[src/examples/libpmemobj/Makefile, line 44 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4ufAUa3D6Lcjhqxzqs:-L4ufAUa3D6Lcjhqxzqt:b-jrxcue) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/examples/libpmemobj/Makefile#L44)):*
> ```Makefile
> 	linkedlist list_map slab_allocator
> 
> ifeq ($(ARCH), x86_64)
> ```

comment

---

*[src/examples/libvmem/Makefile, line 42 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4ufDX4ZgVcAMCupyvy:-L4ufDX4ZgVcAMCupyvz:b-jrxcue) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/examples/libvmem/Makefile#L42)):*
> ```Makefile
> DIRS =
> 
> ifeq ($(ARCH), x86_64)
> ```

comment

---

*[src/libpmem/libpmem.vcxproj, line 34 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4uaSdKxRITzQ1glKYf:-L4uaSdKxRITzQ1glKYg:bev4mj0) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/libpmem/libpmem.vcxproj#L34)):*
> ```XML
>     <ClCompile Include="..\libpmem\libpmem_main.c" />
>     <ClCompile Include="..\windows\win_mmap.c" />
>     <ClCompile Include="..\..\src\libpmem\x86_64\cpu.c" />
> ```

"..\src\" prefix could be removed. Same below.
But I guess this is generated by VS.

---

*[src/libpmem/Makefile, line 59 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4ueFFFsMCkCtOcDMEX:-L4ueFFFsMCkCtOcDMEY:b8mhlh1) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/libpmem/Makefile#L59)):*
> ```Makefile
> ifeq ($(ARCH), x86_64)
>   include x86_64/sources.inc
> else
> ```

```
else
  ifeq ($(ARCH),aarch64)
     ...
  else
$(error unsupported ARCH: $ARCH)
```
Or maybe put an ARCH check into src/Makefile.inc and fail early if it's not x86_64 or aarch64 (same as we do for 32-bit arch).

---

*[src/libpmem/Makefile, line 67 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4uej7MijelqUZFSXgi:-L4uej7NgiRsKb1fTn6o:b-7ropln) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/libpmem/Makefile#L67)):*
> ```Makefile
> ifeq ($(ARCH), x86_64)
>   include x86_64/flags.inc
> else
> ```

ditto

---

*[src/libpmem/pmem_posix.c, line 63 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4uezRkg06Z1VmJvteu:-L4uezRkg06Z1VmJvtev:b7z8abg) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/libpmem/pmem_posix.c#L63)):*
> ```C
> }
> 
> void
> ```

missing brief, missing LOG()

---

*[src/libpmem/pmem_windows.c, line 183 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4uewD-VRz5d8C3tdB_:-L4uewD-VRz5d8C3tdBa:b-fcnw5m) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/libpmem/pmem_windows.c#L183)):*
> ```C
> }
> 
> void
> ```

missing brief, LOG()

---

*[src/test/pmem_movnt_align/pmem_movnt_align.c, line 55 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4ui18KKThe2FdEF14c:-L4ui18KKThe2FdEF14d:b4ozgkt) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/test/pmem_movnt_align/pmem_movnt_align.c#L55)):*
> ```C
> typedef void *(*mem_fn)(void *, const void *, size_t);
> 
> static char *src, *dst;
> ```

each decl in its own line, pls
global var names capitalized, pls

---

*[src/test/unittest/unittest.sh, line 2337 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4ucwsexsyaR6c3k7VL:-L4ucwsexsyaR6c3k7VM:bb3vjeh) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/test/unittest/unittest.sh#L2337)):*
> ```Shell
> 		fi
> 
> 		for N in $NODES_SEQ; do
> ```

any comment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2326)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 2 of 180 files at r9, 5 of 182 files at r10, 13 of 25 files at r11, 2 of 8 files at r12.
Review status: 142 of 223 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[src/common/util.h, line 161 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4ud_Z4t0mO_myJEHfQ:-L5OZ_fhaaeyyhQ0dQR9:b-hvdi9x) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/common/util.h#L161)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

I renamed it to util_is_pow2.

I checked both versions and mine is 8 instructions, while the one using popcount is 28 instructions (with gcc -O3).

</blockquote></details>

interesting...

---

*[src/common/util.h, line 156 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5OZGqU7e3XT2qkfMBa:-L5OZGqU7e3XT2qkfMBb:baho6z1) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/common/util.h#L156)):*
> ```objc
> 
> /*
>  * single_bit_set -- returns !0 when there's only 1 bit set in v, 0 otherwise
> ```

update comment

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2326)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 12 of 180 files at r9, 122 of 182 files at r10, 6 of 25 files at r11, 6 of 8 files at r12.
Review status: 209 of 223 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[src/libpmem/aarch64/init.c, line 125 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5P6jvzFr62tI7LqoQH:-L5P6jvzFr62tI7LqoQI:b-6oleat) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/aarch64/init.c#L125)):*
> ```C
> }
> 
> struct pmem_funcs
> ```

missing brief

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2326)*
<!-- Sent from Reviewable.io -->

gaweinbergi commented:
*[src/common/util.h, line 161 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4ud_Z4t0mO_myJEHfQ:-L5Q-knwpgEjYsedzrv0:b-65ufy5) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/common/util.h#L161)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

interesting...
</blockquote></details>

You should use util_popcount. When compiled with the appropriate -march option, this will compile directly to a single popcnt instruction. Any program not built to enable the popcnt instruction doesn't care about performance in this case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2326)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 11 of 182 files at r10, 6 of 25 files at r11.
Review status: all files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[src/libpmem/pmem.c, line 183 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5TzREYr-a6gQIlsXX9:-L5TzREYr-a6gQIlsXXA:b-87z0eu) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/pmem.c#L183)):*
> ```C
> #include "os_deep_persist.h"
> 
> static struct pmem_funcs funcs;
> ```

Funcs

---

*[src/libpmem/pmem.h, line 69 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U-5vfCHLrXDVkLk36:-L5U-5vfCHLrXDVkLk37:b-c9v4cr) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/pmem.h#L69)):*
> ```objc
> 
> static force_inline void
> noflush(const char *addr)
> ```

what is it for? (comment?)

---

*[src/libpmem/x86_64/avx.h, line 43 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U7vXxvcyMZivpq7iD:-L5U7vXxvcyMZivpq7iE:bmb08j9) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/avx.h#L43)):*
> ```objc
> {
> 	/*
> 	 * _mm256_zeroupper clears upper parts of avx registers.
> ```

Why this comment is inside the function, not above it?

---

*[src/libpmem/x86_64/init.c, line 276 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U57qo8xYFayWXGqKu:-L5U57qo8xYFayWXGqKv:bntkfd2) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L276)):*
> ```C
> 		ASSERT(0);
> #else
> 	LOG(3, "avx supported, but disabled at build time");
> ```

how do you know it's supported?

---

*[src/libpmem/x86_64/init.c, line 316 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U5GNV4RL-6pBVnD4D:-L5U5GNV4RL-6pBVnD4E:b32v1h) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L316)):*
> ```C
> 		ASSERT(0);
> #else
> 	LOG(3, "avx512f supported, but disabled at build time");
> ```

''

---

*[src/libpmem/x86_64/init.c, line 339 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U5Q_seHNafXxwuL0m:-L5U5Q_seHNafXxwuL0n:b-m44t0g) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L339)):*
> ```C
> 		if (e && strcmp(e, "1") == 0)
> 			LOG(3, "PMEM_NO_CLFLUSHOPT forced no clflushopt");
> 		else {
> ```

when using brackets - use it both for if and else

---

*[src/libpmem/x86_64/init.c, line 371 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U5ZIkShmn3Ro8f4bP:-L5U5ZIkShmn3Ro8f4bQ:b-7smsff) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L371)):*
> ```C
> }
> 
> struct pmem_funcs
> ```

brief

---

*[src/libpmem/x86_64/init.c, line 372 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U6ErH68cuWXzdknWD:-L5U6ErH68cuWXzdknWE:bptmfph) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L372)):*
> ```C
> 
> struct pmem_funcs
> pmem_arch_init(void)
> ```

Why this function returns pmem_funcs structure, instead of taking the pointer to the structure that it fills up with func pointers (same as the function above)?

---

*[src/libpmem/x86_64/init.c, line 374 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U5_raSz37IA7C7P6b:-L5U5_raSz37IA7C7P6c:b-85dcqo) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L374)):*
> ```C
> pmem_arch_init(void)
> {
> 	struct pmem_funcs funcs;
> ```

LOG()

---

*[src/libpmem/x86_64/init.c, line 378 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U5cyqXY6Llrv0NJCc:-L5U5cyqXY6Llrv0NJCd:b3c2q0i) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L378)):*
> ```C
> 	funcs.predrain_fence = predrain_fence_empty;
> 	funcs.flush = flush_clflush;
> 	funcs.is_pmem = NULL;
> ```

Why not to set it to *is_pmem_never*, as in the original implementation?

---

*[src/libpmem/x86_64/init.c, line 425 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U6srfcS94undJ9nLF:-L5U6srfcS94undJ9nLG:bqcehsc) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L425)):*
> ```C
> 		funcs.memmove_nodrain == memmove_nodrain_avx512f_clwb ||
> 		funcs.memmove_nodrain == memmove_nodrain_avx512f_empty)
> 		LOG(3, "using movnt AVX512F");
> ```

Isn't it easier to move this LOG() to line 367, instead of having this huge if() here?
I know, the order of debug traces would change, but who cares?

---

*[src/libpmem/x86_64/memcpy_memset.h, line 61 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U7OzSlzC0oSwjKuEn:-L5U7OzSlzC0oSwjKuEo:bb9t4dd) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/memcpy_memset.h#L61)):*
> ```objc
> #ifndef AVX512F_AVAILABLE
> #ifdef _MSC_VER
> #define AVX512F_AVAILABLE 0
> ```

Any XXX comment?

---

*[src/test/util_poolset/util_poolset.vcxproj, line 52 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5UAa6nSxQfvXzLiB8r:-L5UAa6nSxQfvXzLiB8s:b-gen4wj) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/test/util_poolset/util_poolset.vcxproj#L52)):*
> ```XML
>       <AdditionalIncludeDirectories>$(SolutionDir)\libpmem;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
>       <ForcedIncludeFiles>mocks_windows.h;%(ForcedIncludeFiles)</ForcedIncludeFiles>
>       <PreprocessorDefinitions>PMDK_UTF8_API;NTDDI_VERSION=NTDDI_WIN10_RS1;SSE2_AVAILABLE=0;AVX_AVAILABLE=0;AVX512F_AVAILABLE=0;_CONSOLE;%(PreprocessorDefinitions)</PreprocessorDefinitions>
> ```

Most of this macros are defined in common property sheet, so you should inherit them and add only the additional ones. Something like:

`<PreprocessorDefinitions>SSE2_AVAILABLE=0;AVX_AVAILABLE=0;AVX512F_AVAILABLE=0;%(PreprocessorDefinitions)</PreprocessorDefinitions>`

---

*[src/test/util_poolset/util_poolset.vcxproj, line 60 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5UEFwjP11LMFI3_Ur7:-L5UEFwjP11LMFI3_Ur8:b32v1h) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/test/util_poolset/util_poolset.vcxproj#L60)):*
> ```XML
>       <AdditionalIncludeDirectories>$(SolutionDir)\libpmem;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
>       <ForcedIncludeFiles>mocks_windows.h;%(ForcedIncludeFiles)</ForcedIncludeFiles>
>       <PreprocessorDefinitions>PMDK_UTF8_API;NTDDI_VERSION=NTDDI_WIN10_RS1;SSE2_AVAILABLE=0;AVX_AVAILABLE=0;AVX512F_AVAILABLE=0;_CONSOLE;%(PreprocessorDefinitions)</PreprocessorDefinitions>
> ```

''

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2326)*
<!-- Sent from Reviewable.io -->

krzycz commented:
General comment - missing briefs for many functions, or at least some brief at the beginning of the file.

---

Review status: all files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2326#-:-L5UEqUOFxoe4XMh41PQ:b-kuxwr5)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Review status: 200 of 223 files reviewed at latest revision, 21 unresolved discussions.

---

*[src/common/util.h, line 161 at r10](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L4ud_Z4t0mO_myJEHfQ:-L5UUAqCsckVi2pMtob8:b9jvndu) ([raw file](https://github.com/pmem/pmdk/blob/242b7058fedb89ea33c2b83b6b0b1fbac33e596e/src/common/util.h#L161)):*
<details><summary><i>Previously, gaweinbergi (Glenn Weinberg) wrote…</i></summary><blockquote>

You should use util_popcount. When compiled with the appropriate -march option, this will compile directly to a single popcnt instruction. Any program not built to enable the popcnt instruction doesn't care about performance in this case.
</blockquote></details>

No distribution compiles with -march= (with an exception of Gentoo :)... And we don't really need number of bits set, just whether there is exactly one.

---

*[src/common/util.h, line 156 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5OZGqU7e3XT2qkfMBa:-L5UU7qkB5-2ux97I5wA:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/common/util.h#L156)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

update comment
</blockquote></details>

Done.

---

*[src/libpmem/pmem.c, line 183 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5TzREYr-a6gQIlsXX9:-L5UVCJcI296tvX8nhOs:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/pmem.c#L183)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Funcs
</blockquote></details>

Done.

---

*[src/libpmem/pmem.h, line 69 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U-5vfCHLrXDVkLk36:-L5UXfYgS0eYYgee7X1H:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/pmem.h#L69)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

what is it for? (comment?)
</blockquote></details>

Done.

---

*[src/libpmem/aarch64/init.c, line 125 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5P6jvzFr62tI7LqoQH:-L5UXi5ppDrUcOrRNkJS:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/aarch64/init.c#L125)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

missing brief
</blockquote></details>

Done.

---

*[src/libpmem/x86_64/avx.h, line 43 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U7vXxvcyMZivpq7iD:-L5UYNKfIdiek7-W9dIA:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/avx.h#L43)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Why this comment is inside the function, not above it?
</blockquote></details>

Done.

---

*[src/libpmem/x86_64/init.c, line 276 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U57qo8xYFayWXGqKu:-L5UYiodJuSTQmVVmisv:b74y845) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L276)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

how do you know it's supported?
</blockquote></details>

Because this function is called only when AVX is supported.

---

*[src/libpmem/x86_64/init.c, line 339 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U5Q_seHNafXxwuL0m:-L5UZMpF4Bs5BwuRKDOj:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L339)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

when using brackets - use it both for if and else
</blockquote></details>

Done.

---

*[src/libpmem/x86_64/init.c, line 371 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U5ZIkShmn3Ro8f4bP:-L5UZNGF00bvuCgU7ZC6:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L371)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

brief
</blockquote></details>

Done.

---

*[src/libpmem/x86_64/init.c, line 372 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U6ErH68cuWXzdknWD:-L5UZYU1rBp69qceMhKa:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L372)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Why this function returns pmem_funcs structure, instead of taking the pointer to the structure that it fills up with func pointers (same as the function above)?
</blockquote></details>

Done.

---

*[src/libpmem/x86_64/init.c, line 374 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U5_raSz37IA7C7P6b:-L5UZXOD76VqIFZqt6XX:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L374)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

LOG()
</blockquote></details>

Done.

---

*[src/libpmem/x86_64/init.c, line 378 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U5cyqXY6Llrv0NJCc:-L5U_12_H8QlMYwiXfRS:b-xf5wdd) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L378)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Why not to set it to *is_pmem_never*, as in the original implementation?
</blockquote></details>

Because is_pmem_never is in other module now.

---

*[src/libpmem/x86_64/init.c, line 425 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U6srfcS94undJ9nLF:-L5UpXu2VGalp3_i6A79:bj4dpim) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/init.c#L425)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Isn't it easier to move this LOG() to line 367, instead of having this huge if() here?
I know, the order of debug traces would change, but who cares?
</blockquote></details>

I simplified this a bit. Are you satisfied now?

---

*[src/libpmem/x86_64/memcpy_memset.h, line 61 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5U7OzSlzC0oSwjKuEn:-L5UaRGH_4k_ETmko1nZ:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/libpmem/x86_64/memcpy_memset.h#L61)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Any XXX comment?
</blockquote></details>

Done.

---

*[src/test/util_poolset/util_poolset.vcxproj, line 52 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5UAa6nSxQfvXzLiB8r:-L5Uai-9PoN6pZkMBzRy:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/test/util_poolset/util_poolset.vcxproj#L52)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Most of this macros are defined in common property sheet, so you should inherit them and add only the additional ones. Something like:

`<PreprocessorDefinitions>SSE2_AVAILABLE=0;AVX_AVAILABLE=0;AVX512F_AVAILABLE=0;%(PreprocessorDefinitions)</PreprocessorDefinitions>`




</blockquote></details>

Done.

---

*[src/test/util_poolset/util_poolset.vcxproj, line 60 at r12](https://reviewable.io:443/reviews/pmem/pmdk/2326#-L5UEFwjP11LMFI3_Ur7:-L5UaiWVFa0kxaNXfxHE:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/f943011a221780e0f4cde2955139dbbdad12d39c/src/test/util_poolset/util_poolset.vcxproj#L60)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

''
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2326)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 7 of 182 files at r10, 1 of 25 files at r11, 21 of 23 files at r13.
Review status: 221 of 223 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2326)*
<!-- Sent from Reviewable.io -->

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 23 files at r13.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2326#-:-L5hYDoZwFRckbFh4UJG:bnfp4nl)*
<!-- Sent from Reviewable.io -->
